### PR TITLE
App signing

### PR DESCRIPTION
**Pull request by LLFourn**

### CODE REVIEW

File: coordinator-cli/src/signer.rs [hunk 1]
```diff
@@ -147,18 +132,13 @@ impl<'a, 'b> Signer<'a, 'b> {
             // because often a big bunch of devices will register at similar times if they are daisy
             // chained together.
             loop {
+                std::thread::sleep(Duration::from_millis(100));
```
Comment (nickfarrow) [inline coordinator-cli/src/signer.rs:135]: What's the motivation behind this sleep?

File: frostsnap_coordinator/src/signing.rs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+use std::collections::HashSet;
+
+pub struct SigningDispatcher {
+    request: SignRequest,
+    targets: HashSet<DeviceId>,
+    got_signatures: HashSet<DeviceId>,
+    need_to_send_to: HashSet<DeviceId>,
+}
+
+impl SigningDispatcher {
```
Comment (nickfarrow) [inline frostsnap_coordinator/src/signing.rs:17]: LGTM

**Comment by nickfarrow** [comment]
Looked over recent changes, going to test now.

One question I have is whether this `state_counter` pattern is going to come up elsewhere? E.g. could we have similar race conditions for a `wallet transaction history` display syncing with the blockchain?

**Review by nickfarrow — APPROVED** [review]

File: coordinator-cli/src/signer.rs (hunk 1 above)
Comment (LLFourn) [reply coordinator-cli/src/signer.rs:135]: This loop hoses CPU otherwise. It should have a comment.

**Comment by LLFourn** [comment]
> Looked over recent changes, going to test now.
> 
> One question I have is whether this `state_counter` pattern is going to come up elsewhere? E.g. could we have similar race conditions for a `wallet transaction history` display syncing with the blockchain?

I think in theory you can have these race conditions when you start streams in this way but they are just very unlikely. Doing things this way seems like good practice as it helps you recover from panics that occur during processing the event (this was the actual problem) by rebuilding the list.

**Pull request merged**